### DISCUSSION TITLE
Why did the fine tuned model degrade previous learnt voices ?

### DISCUSSION THREAD

**Discussion by Ca-ressemble-a-du-fake** [coqui-ai/TTS#2069] [General Q&A]
Hi,

I have trained a multi speaker VITS model with 4 speakers for 100 k iterations. Results sound good to my ears. Now I have a dataset with only 0.5k samples (~ 1/10th of each previous speaker dataset) so I fine tuned the previous 4 speaker model with frozen encoder and duration predictor for 30 k steps which improved the quality of the target (with fewer samples) speaker but much degraded the previous learnt voices (they now sound metallic and un-intelligible). Why did that happen ? I thought freezing the encoder and duration predictor would make the initial voices remain as they were but it is not the case. What should I do to keep the previous voices with high quality ?

Here is my process :
1. Train model with 4 datasets after computing the embeddings and providing the d vector file to the model.
2. Compute embeddings of the 4 previous speakers + the new one and provide the file to the model.
3. Fine tune the previous 4 voice model (with frozen encoder and duration predictor) with only the new dataset (as I understood they did in the paper YourTTS with Experiment 4)
4. When I `synthetize.py --text "This is a test bla bla." --model_path myFinedTunedModel --config_path myFinedTunedModel_config.json --speaker_idx FIRST_VOICE --out_path herePath` the output sounds metalic and of poor quality

Thanks for your help

**Comment by iamkhalidbashir** [comment]
Interested in this. Did you find a solution to why the previous degraded? is it because you only new dataset?
Also Should I fine-tune with freezing or without freezing?

Thanks

**Reply by Ca-ressemble-a-du-fake** [reply]
Hi @iamkhalidbashir ! Actually from what I understood it is "better" to dedicate all the model neurons to a single voice. So now I am training every single voice/speaker from scratch. Freezing anything did not yield good results for me. Maybe I will try finetuning again later.

**Reply by iamkhalidbashir** [reply]
So what do you do if you have less data for a voice ?  You get more data and dosent fine tune ?
Unfortunately for my case i need fine tuning

i might try weighted sampling like in the yourtts paper for fine tuning

**Reply by Ca-ressemble-a-du-fake** [reply]
Yeah thats's right I am spending lots of time creating a dataset as large as I can ! Did you try the recently released recipe for yourtts (it was released around last December)?
Or maybe did you give IMS Toucan (this is a model available on github) a try ?

**Comment by iamkhalidbashir** [comment]
Yes I currently training on it (yourtts), will update if it works great
IMS Toucan looks promising I think it uses fast speech 2 but I couldnt find any voice clone colab ;/

**Reply by Ca-ressemble-a-du-fake** [reply]
Indeed it uses FS2. You mean there is no colab notebook available ? The instructions on the site are very straightforward you shouldn't need more. Anyway looking forward to reading your update ! By the way who are you trying to reproduce and in which language ?

**Reply by AlexSteveChungAlvarez** [reply]
I have trained on IMS Toucan with the same datasets as I commented below, it maintains languages when fine tuning, but it needs at least 2 languages, if not, the accent of the only language you fine tune influences the others. It doesn't clone at all the voice of the reference speaker (not seen during training), it only uses the voice it's trained with. But as you may hear on my results, yourtts doesn't too. Toucan only needs a few steps to train (took me about 1 hour to fine tune on ~7k steps) and get good results.

**Comment by AlexSteveChungAlvarez** [comment]
Hello, I have tried to fine tune YourTTS model with a new language, but it hasn't given me the results I expected:

1. When I fine tune on the released model, the languages disappear and only the language(s) of the dataset(s) that I use for training remain, as if I were training from scratch. I have tried fine tuning with  only 1 new language first and then with 2 languages, getting the same result.
2. For the first language (Quechua) I use a single speaker dataset of a few minutes of speech, the second language is English and I use libritts 360. When I fine tune with the first dataset only, it gives me good results at ~7k steps when synthethizing speech from the seen speaker during training, whereas when I try to clone another voice, it doesn't. When I fine tune with both datasets for ~50k steps speech is not well synthesised for any of the languages and the speakers don't sound as the reference speakers too. I attach the reference audios and the generated ones:
[references.zip](https://github.com/coqui-ai/TTS/files/10423276/references.zip)
[results.zip](https://github.com/coqui-ai/TTS/files/10423236/results.zip)

**Reply by AlexSteveChungAlvarez** [reply]
@Edresson can you give me some advice?

**Reply by Edresson** [reply]
> 1. he released model, the languages disappear and only the language(s) of the dataset(s) that I use for training remain, as if I were trainin

The model should not keep the others languages if these languages are not included in the training set. So the first is not a bug but expected behavior.

About the second, You might have a bug in your training recipe, because it should work if you are training enough. Please could you share the training recipe that you are using?

**Reply by AlexSteveChungAlvarez** [reply]
I expected the languages to be kept because I am trying also the Toucan Toolkit and when finetuning, with the same 2 datasets, it does keep the pretrained languages. That's why I thought YourTTS would do it, too.
I attach the recipes for single speaker quechua training and for single speaker quechua+libritts.
[recipes_quechua.zip](https://github.com/coqui-ai/TTS/files/10428737/recipes_quechua.zip)

**Reply by Edresson** [reply]
Hey you should define the vocabulary and phonemes from English + your extra language as we did here: https://github.com/coqui-ai/TTS/blob/6e3f74fc2927892b5575d33e4df757e6d9f19fd3/recipes/vctk/yourtts/train_yourtts.py#L173

In addition, you should follow the Yourtts recipe and not change parameters until your fine-tuning works. I saw that you are using phonemes and I think you should disable it and train it on the char level first to guarantee that it is not the issue.

**Reply by AlexSteveChungAlvarez** [reply]
where did you define phonemes? I see in the code `phonemes: ""`. 
I didn't train on the char level first because I can make a mistake in the vocabulary, I would look for it on the internet as I don't speak Quechua. Besides, I see you put in the english vocabulary some Unicodes, which I don't know what they are for, would you tell me please? I see in the characters the only letter that's left for quechua would be the 'ñ', as for spanish too. But I don't know about the unicodes...

**Reply by AlexSteveChungAlvarez** [reply]
I already trained on graphemes over 58k steps.
[results.zip](https://github.com/coqui-ai/TTS/files/10452644/results.zip)
I attach the new results + the recipe, what do you think? @Edresson

**Reply by Edresson** [reply]
> where did you define phonemes? I see in the code `phonemes: ""`. I didn't train on the char level first because I can make a mistake in the vocabulary, I would look for it on the internet as I don't speak Quechua. Besides, I see you put in the english vocabulary some Unicodes, which I don't know what they are for, would you tell me please? I see in the characters the only letter that's left for quechua would be the 'ñ', as for spanish too. But I don't know about the unicodes...


To get the characters vocabulary you can use [this script](https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/find_unique_chars.py). You just need to provide the config.json and it will return the vocabulary. 

About the unicodes it is just a encode problem that happens when the file was save. Python will deal with no problem and will consider these characters as something like ""ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz¯·ßàáâãäæçèéêëìíîïñòóôõöùúûüÿāąćēęěīıłńōőœśūűźżǎǐǒǔабвгдежзийклмнопрстуфхцчшщъыьэюяёєіїґ–!'(),-.:;? ".

About phonemes YourTTS is trained with  graphemes then I kept it empty.  The default phonemes already have almost all phonemes on IPA and you can found it here: https://github.com/coqui-ai/TTS/blob/dev/TTS/tts/utils/text/characters.py#L27-L35 (they are automattically used when you dont provide the CharactersConfig class). You can  also get all the phonemes present in your dataset with the script: https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/find_unique_phonemes.py


 

> I already trained on graphemes over 58k steps. [results.zip](https://github.com/coqui-ai/TTS/files/10452644/results.zip) I attach the new results + the recipe, what do you think? @Edresson



Ok looking for your recipe, you are using speaker batching instead of language batch then the training will not be balanced by languages.
You should set `weighted_sampler_attrs={"language": 1.0},`. Your audios looks very weird, please check if yours datasets  sampling rate are 16000. If not you can use the [resample script](https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/resample.py) to resample it. In addition, depending on your data, might be interesting trimming the begging and end utterances silences with [this scrip](https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/remove_silence_using_vad.py).

**Reply by AlexSteveChungAlvarez** [reply]
> About the unicodes it is just a encode problem that happens when the file was save. Python will deal with no problem and will consider these characters as something like ""ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz¯·ßàáâãäæçèéêëìíîïñòóôõöùúûüÿāąćēęěīıłńōőœśūűźżǎǐǒǔабвгдежзийклмнопрстуфхцчшщъыьэюяёєіїґ–!'(),-.:;? ".

Thank you for the explanation!

>About phonemes YourTTS is trained with graphemes then I kept it empty. The default phonemes already have almost all phonemes on IPA and you can found it here: https://github.com/coqui-ai/TTS/blob/dev/TTS/tts/utils/text/characters.py#L27-L35 (they are automattically used when you dont provide the CharactersConfig class). You can also get all the phonemes present in your dataset with the script: https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/find_unique_phonemes.py

Nice! I didn't know about that script for getting the phonemes.

> Your audios looks very weird, please check if yours datasets sampling rate are 16000.

I resampled it before training, yes, and LibriTTS already comes in that sampling rate.

> Ok looking for your recipe, you are using speaker batching instead of language batch then the training will not be balanced by languages.

Yes! I didn't know how to use language batching to balance it, since libritts is a lot bigger than the dataset I am using, it should be as you say.

> You should set weighted_sampler_attrs={"language": 1.0},. Your audios looks very weird, 

I will try this and come up with new results in a couple of days, thank you very much. Yes, the audios sound very weird. Unfortunately, I had to present in the document for my last course at university the results I had as they were in this last zip I sent you, but I will try to present the updated results in the oral presentation. About the trimming, the audios on my data are about 2-3 seconds long, I don't think they need to be trimmed!

**Reply by Edresson** [reply]
> > About the unicodes it is just a encode problem that happens when the file was save. Python will deal with no problem and will consider these characters as something like ""ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz¯·ßàáâãäæçèéêëìíîïñòóôõöùúûüÿāąćēęěīıłńōőœśūűźżǎǐǒǔабвгдежзийклмнопрстуфхцчшщъыьэюяёєіїґ–!'(),-.:;? ".
> Thank you for the explanation!
> >About phonemes YourTTS is trained with graphemes then I kept it empty. The default phonemes already have almost all phonemes on IPA and you can found it here: https://github.com/coqui-ai/TTS/blob/dev/TTS/tts/utils/text/characters.py#L27-L35 (they are automattically used when you dont provide the CharactersConfig class). You can also get all the phonemes present in your dataset with the script: https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/find_unique_phonemes.py
> Nice! I didn't know about that script for getting the phonemes.
> > Your audios looks very weird, please check if yours datasets sampling rate are 16000.
> I resampled it before training, yes, and LibriTTS already comes in that sampling rate.
> 
> > Ok looking for your recipe, you are using speaker batching instead of language batch then the training will not be balanced by languages.
> Yes! I didn't know how to use language batching to balance it, since libritts is a lot bigger than the dataset I am using, it should be as you say.
> 
> > You should set weighted_sampler_attrs={"language": 1.0},. Your audios looks very weird, 
> I will try this and come up with new results in a couple of days, thank you very much. Yes, the audios sound very weird. Unfortunately, I had to present in the document for my last course at university the results I had as they were in this last zip I sent you, but I will try to present the updated results in the oral presentation. About the trimming, the audios on my data are about 2-3 seconds long, I don't think they need to be trimmed!



LibriTTS is 22khz not 16khz. Then check it again.

**Reply by AlexSteveChungAlvarez** [reply]
You are right! I don't know where did I read it was 16khz, I just read it again and it says 24khz. I am resampling it right now.

> To get the characters vocabulary you can use [this script](https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/find_unique_chars.py). You just need to provide the config.json and it will return the vocabulary.

About this, which config.json do I need to provide? What does this file need to include inside?

**Reply by Edresson** [reply]
> You are right! I don't know where did I read it was 16khz, I just read it again and it says 24khz. I am resampling it right now.
> 
> > To get the characters vocabulary you can use [this script](https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/find_unique_chars.py). You just need to provide the config.json and it will return the vocabulary.
> 
> About this, which config.json do I need to provide? What does this file need to include inside?
> 

It need to have your datasets configs. Use the config from your previous training.

**Reply by AlexSteveChungAlvarez** [reply]
> To get the characters vocabulary you can use [this script](https://github.com/coqui-ai/TTS/blob/dev/TTS/bin/find_unique_chars.py). You just need to provide the config.json and it will return the vocabulary.

 I got this:
` > Number of unique characters: 77
 > Unique characters:
 !"'(),-./:;?ABCDEFGHIJKLMNOPQRSTUVWXYZ[]abcdefghijklmnopqrstuvwxyz¯½¿æéêïœ—
 > Unique lower characters: abcdefghijklmnopqrstuvwxyzæéêïœ
 > Unique all forced to lower characters:
 !"'(),-./:;?[]abcdefghijklmnopqrstuvwxyz¯½¿æéêïœ—`

However, I don't see the "Ñ" and "ñ". Is this OK? Should I put this result into the character config? Or are these the characters that are already by default?
 I already changed in the recipe speaker batching for language batching and resampled Libritts to 16 khz. I will wait for your confirmation to start training, thank you for guiding me!
